File: miotapi.py
import requests
import urllib.parse
import os.path
import inspect
import sys
from datetime import datetime


class AttributedDict(dict): 
    def __getattr__(self, item):
        return self[item]


class User:
    def __init__(self, api, data):
        self.api = api
        data = AttributedDict(data)
        self.uuid = data.uuid
        self.telegram_id = int(data.telegram_id)
        self.telegram_username = data.get('telegram_username', None)
        self.name = data.name
        self.flags = data.flags
        self.registration_date = datetime.strptime(data.registration_date, '%Y-%m-%d %H:%M:%S.%f')

    def profile_page(self):
        return f'{self.api.protocol}://{self.api.host}/{self.uuid}'
    
    def __repr__(self):
        return f'\nUser {self.uuid}\n'\
               f'Name: {self.name}\n'\
               f'Telegram ID: {self.telegram_id}\n'\
               f'Telegram username: @{self.telegram_username}\n'\
               f'Registration date: {self.registration_date}\n'\
               f'Flags: {", ".join(self.flags)}\n'


class AssetsManager:
    @staticmethod
    def get_root_path():
        return os.path.dirname(inspect.getfile(sys.modules[__name__]))
    
    @staticmethod
    def get_assets_path():
        return AssetsManager.get_root_path() + '/assets'
    
    @staticmethod
    def get_items_path():
        return AssetsManager.get_assets_path() + '/items'
    
    @staticmethod
    def get_item_path(item_id):
        items_path = AssetsManager.get_items_path()
        item_path = items_path + '/' + item_id
        if os.path.isfile(item_path + '.webp'):
            return item_path + '.webp'
        if os.path.isfile(item_path + '.png'):
            return item_path + '.png'


class MiotAPI:
    VERSION = '0.1.1'

    @staticmethod
    def dict_to_args(d):
        args = []
        for key in d.keys():
            if d[key] is not None:
                args.append(f'{key}={urllib.parse.quote_plus(str(d[key]))}')
        return '?' + '&'.join(args) if len(args) > 0 else ''


    def __init__(self, host: str, protocol: str = 'https', prefix: str = '/api', timeout=1):
        self.host = host
        self.protocol = protocol
        self.prefix = prefix
        self.timeout = timeout
    
    def _status(self):
        try:
            res = self.get('/')
            return res['status']
        except Exception:
            return 'offline'
    
    def get(self, meth, tries=3):
        try:
            return requests.get(f'{self.protocol}://{self.host}{self.prefix}{meth}', timeout=self.timeout).json()
        except Exception as e: 
            if tries > 1:
                return self.get(meth, tries-1)
            raise e

    def get_users(self, uuid: str=None, telegram_id: int=None, telegram_username: str=None, name: str=None, flags: list[str]=None,
                  registration_date: int=None, roles: list[str]=None) -> list[AttributedDict]:
        args = self.dict_to_args({
             'uuid': uuid,
             'telegram_id': telegram_id,
             'telegram_username': telegram_username,
             'name': name,
             'flags': ','.join(flags) if flags is not None else None,
             'registration_date': registration_date,
             'roles': ','.join(roles)
if roles is not None else None
        })
        return [
             User(self, user) 
             for user in self.get(f'/users{args}')
        ]
    
    def search_users(self, query: str) -> list[AttributedDict]:
        return [
             User(self, user) 
             for user in self.get(f'/search/users?query={urllib.parse.quote_plus(query)}')
        ]
         
    def get_search_suggestions(self, query: str) -> list[str]:
        return self.get(f'/search.suggestions?query={urllib.parse.quote_plus(query)}')['suggestions']
         

if __name__ == '__main__':
    print(AssetsManager.items_path())
    print(AssetsManager.item_path('miotbot'))
    api = MiotAPI('mrxx.ru')
    print(api._status())
    print(api.get_users(uuid='u/mrybs'))
    print()
    print(api.get_search_suggestions('чин'))
    print()
    print(api.search_users('чинчопа <3'))
    print()
    print(api.get_users(telegram_id=960063512)[0].uuid)

File: miotapi.cpp
#pragma once

#include <string>
#include <list>
#include <curl/curl.h>
#include <iostream>
#include "json.hpp"

namespace Miot{
    static const char* VERSION = "0.1.2";

    class User{
        const void* api;
        const nlohmann::json& data;

    public:
        const std::string uuid;
        const int telegram_id;
        const std::string telegram_username;
        const std::string name;
        const std::vector<std::string>& flags;
        const std::string registration_date;


        User(const void* api, const nlohmann::json& data)
            :api(api), data(data), 
            uuid(std::string(data["uuid"])),
            telegram_id(data["telegram_id"]),
            telegram_username(data.contains("telegram_username") ? std::string(data["telegram_username"]) : nullptr),
            name(std::string(data["name"])),
            flags(data["flags"]),
            registration_date(std::string(data["registration_date"])){}

        operator std::string() const{
            std::stringbuf buffer;
            std::ostream stream(&buffer);
            stream << "User " << uuid;
            stream << "\nName: " << telegram_id;
            stream << "\nTelegram ID: " << telegram_id;
            stream << "\nTelegram username: @" << telegram_username;
            stream << "\nRegistration date: " << registration_date;
            return buffer.str();
        }
    };

    class API{
        const char* host;
        const char* protocol;
        const char* prefix;
        const unsigned long timeout;
        CURL* curl;
    
    protected:
        inline static size_t writeCallback(void* contents, size_t size, size_t nmemb, void* userp) {
            ((std::string*)userp)->append((char*)contents, size * nmemb);
            return size * nmemb;
        }

    private:
        static std::string makeHexByChar(char Char){
            std::stringstream buffer;
            buffer << std::hex << int(Char);
            return buffer.str();
        }

    protected:
        static std::string URLEncode(const std::string& text){
            std::string encoded = std::string();
            for(char c: text){
                encoded += "%";
                if(makeHexByChar(c).size() == 8){
                encoded += makeHexByChar(c)[6];
                encoded += makeHexByChar(c)[7];
                }else
                encoded += makeHexByChar(c);
            }
            return encoded;
        }
        
        static std::string jsonToGetArgs(nlohmann::json j){
            std::list<const char*> _args;
            for(const auto& el : j.items()){
                if(el.value() == nullptr) continue;
                std::cout << el.value() << std::endl;
                std::string arg;
                arg.reserve(el.key().size() + 1 + el.value().size());
                arg.append(el.key()).append("=").append(URLEncode(el.value()));
                _args.push_back(arg.c_str());
            }
            std::string args = "?";
            auto i = 0;
            for(const auto& arg : _args){
                args += arg;
                i++;
                if(i < _args.size()) args += '&';
            }
            return args;
        }

    public:
        API(const char* host, const char* protocol="https", const char* prefix="/api", const unsigned long timeout=1L)
            :host(host), protocol(protocol), prefix(prefix), timeout(timeout){
                curl_global_init(CURL_GLOBAL_DEFAULT);
                curl = curl_easy_init();
                if(!curl){
                    std::cout << "!curl" << std::endl;
                }
            }
        
        inline ~API() noexcept{
            curl_easy_cleanup(curl);
        }
    
    protected:
        nlohmann::json get(const char* meth, const unsigned char tries=3){
            std::string readBuffer;
            std::string url;
            url.reserve(strlen(protocol) + 3 + strlen(host) + strlen(prefix) + strlen(meth));
            url
                .append(protocol)
                .append("://")
                .append(host)
                .append(prefix)
                .append(meth);
            curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
            curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, timeout);
            curl_easy_setopt(curl, CURLOPT_TIMEOUT, timeout);
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
            const auto res = curl_easy_perform(curl);
            if(res != CURLE_OK){
                if(tries > 1) return get(meth, tries - 1);
                std::cerr << "\n\n\u001b[31mcURL error: " << curl_easy_strerror(res) << "\u001b[0m" << std::endl;
                exit(2);
            }
            const auto result = nlohmann::json::parse(readBuffer);
            return result;
        }
    
    public:
        inline const char* status(){
            return std::string(get("/")["status"]).c_str();
        }

        std::vector<User> getUsers(const nlohmann::json& args){
            auto _users = get((std::string("/users") + jsonToGetArgs(args)).c_str());
            std::vector<User> users;
            for(const auto user : _users){
                users.push_back(User(this, nlohmann::json(user)));
            }
            return users;
        }
    };
}
